### PR TITLE
Replace Clipper1 library by Clipper2 library

### LINKED ISSUE #90103: `Geometry2D` polygon boolean operations still use old Clipper1 library
### Tested versions

4.3dev

### System information

Windows 10

### Issue description

I create this issue so the upgrade does not get forgotten again.

The `Geometry2D` class (and the `Sprite2D editor` as the only other user for that matter) still use older `Clipper1` library for the polygon operations.

`Clipper2` is part of Godot since version 4.2 as it was needed for the 2D navigation mesh baking.

According to the developer (and own testing) there is no reason to not upgrade:

- The code is cleaner and faster
- Clipper2 is easier to use with simple functions that handle all common uses
- Floating point coordinates no longer need to be converted to integers before passing them into the library
- And micro self-intersections and polygons with touching edges are now rarely seen in solutions

Clipper2 is a lot faster and scales better with complexity than Clipper1 which means runtime Geometry2D ops are less likely to cause performance issues and have less errors.

clipper_performance

### Steps to reproduce

Replace Clipper1 code found in `core/math/geometry_2d.cpp` with Clipper2 functions and look out for regressions in the results.

The functions that need the upgrade are `_polypaths_do_operation()` and `_polypath_offset()`.

### Minimal reproduction project (MRP)

N/A

### PR DESCRIPTION
**Pull request by rburing**
Reduces release template binary size (by 54.1 kB on Linux).

To-do:

- [x] Resolve TODOs in code
- [x] Use floating point coordinates directly
- [x] Test more extensively

Fixes https://github.com/godotengine/godot/issues/90103

### CODE REVIEW

File: tests/core/math/test_geometry_2d.h [hunk 1]
```diff
@@ -685,12 +685,12 @@ TEST_CASE("[Geometry2D] Clip polyline with polygon") {
...
-		CHECK(r[0][0].is_equal_approx(Vector2(160, 320)));
+		CHECK(r[0][0].is_equal_approx(Vector2(121.412682, 225.038757)));
 		CHECK(r[0][1].is_equal_approx(Vector2(122, 250)));
-		CHECK(r[0][2].is_equal_approx(Vector2(121.412682, 225.038757)));
+		CHECK(r[0][2].is_equal_approx(Vector2(160, 320)));
 		REQUIRE_MESSAGE(r[1].size() == 2, "The resulting clipped line should have 2 vertices.");
-		CHECK(r[1][0].is_equal_approx(Vector2(53.07737, 116.143021)));
-		CHECK(r[1][1].is_equal_approx(Vector2(55, 70)));
+		CHECK(r[1][0].is_equal_approx(Vector2(55, 70)));
+		CHECK(r[1][1].is_equal_approx(Vector2(53.07737, 116.143021)));
```
Comment (rburing) [inline tests/core/math/test_geometry_2d.h:693]: The orientation should be fixed instead of rigging the tests.

**Comment by smix8** [comment]
You can replace the Clipper XYZ_64 classes with the Clipper D class equivalents. Those do the same except that they allow for direct float input that gets up- and downscaled to the integer grid internally. 

So the old Clipper1 manual scaling with SCALE_FACTOR can be removed.

I did a similar upgrade for the 2D navigation mesh baking as an example with PR https://github.com/godotengine/godot/pull/89929 here.

File: tests/core/math/test_geometry_2d.h (hunk 1 above)
Comment (smix8) [reply tests/core/math/test_geometry_2d.h:693]: See https://github.com/godotengine/godot/issues/49716

**Comment by ghmart** [comment]
What's about #29886?

**Comment by smix8** [comment]
> What's about #29886?

This PR should imo stay focus on only the upgrade so it can be added without much issues and bike-shed.

Adding new parameters is feature territory as it creates compatibility issues.

It is likely that there is need for a more general rework of the Geometry2D boolean ops later which would be a good place for this and other features. E.g. the single polygon per function limit comes to mind. It is kinda arbitrary as Clipper2 can do way more.

File: tests/core/math/test_geometry_2d.h (hunk 1 above)
Comment (rburing) [reply tests/core/math/test_geometry_2d.h:693]: I'm not sure how to fix this (i.e. keep compatibility) without adding a hack in `clip_polyline_with_polygon`.

**Comment by rburing** [comment]
@smix8 Thanks for the tips! I used the D variants of Clipper functions/types now, and used `RectClip`.

I could not remove the `SCALE_FACTOR` from `Geometry2D` since the unit tests expect higher precision.

Should we actually be using `Point<real_t>` instead of `PointD` etc.?

File: tests/core/math/test_geometry_2d.h (hunk 1 above)
Comment (rburing) [reply tests/core/math/test_geometry_2d.h:693]: Looking at the picture, I think this is actually an improvement/bugfix. The original polyline goes from top to bottom, and now the clipped lines also go from top to bottom.

**Review by smix8 — APPROVED** [review]
Looks good for the Clipper2 replacements.

Not really a 2D and sprite editor user so can't say if there are some subtle differences in the results that could cause some compatibility problems for older projects.

**Comment by KoBeWi** [comment]
I'm using Geometry2D for merging polygons and this change makes my tool a few times slower. Here's a test scene:
[GeometryTest.tscn.txt](https://github.com/godotengine/godot/files/14904892/GeometryTest.tscn.txt)

This was already discussed in the chat, I thought I'd mention it here for the record.

File: core/math/geometry_2d.cpp [hunk 2]
```diff
@@ -196,126 +196,120 @@ void Geometry2D::make_atlas(const Vector<Size2i> &p_rects, Vector<Point2i> &r_re
...
-	Paths paths;
-	co.Execute(paths, p_delta * (real_t)SCALE_FACTOR); // Inflate/deflate.
+	// Inflate/deflate.
+	PathsD paths = InflatePaths({ polypath }, p_delta, jt, et, 2.0, PRECISION, 0.0);
+	// Here the miter_limit = 2.0 and arc_tolerance = 0.0 are Clipper2 defaults,
+	// and the PRECISION is used to scale points up internally, to attain the desired precision.
 
-	// Have to scale points down now.
 	Vector<Vector<Point2>> polypaths;
+	for (Paths64::size_type i = 0; i < paths.size(); ++i) {
```
Comment (AThousandShips) [inline core/math/geometry_2d.cpp:305]:
```suggestion
	for (PathsD::size_type i = 0; i < paths.size(); ++i) {
```

**Comment by rburing** [comment]
Here is an image that explains why it was slower:

clipper2-collinear

I have now added

```cpp
clp.PreserveCollinear(false); // Remove redundant vertices.
```

to make the resulting polygon the same as it was with Clipper1 (i.e. having just the corner vertices).

Performance is still a bit lower than with Clipper1, but the difference is no longer orders of magnitude.

**Review by akien-mga — APPROVED** [review]

**Pull request merged**